extra/DEVELOPER_DOCS/Code Conventions.md

```diff
@@ -0,0 +1,546 @@
+
+### Avoid mutable default arguments
+
+A common Python gotcha are [mutable default arguments](https://docs.python-guide.org/writing/gotchas/#mutable-default-arguments): if your argument defines a mutable default value like `[]` or `{}` and then goes and mutates it, the default value is created _once_ when the function is created and the same object is then mutated every time the function is called. This can be pretty unintuitive when you first encounter it. We therefore avoid writing logic that does this.
+
+If your arguments need to default to an empty list or dict, you can use the `SimpleFrozenList` and `SimpleFrozenDict` helpers provided by spaCy. They are simple frozen implementations that raise an error if they're being mutated to prevent bugs and logic that accidentally mutates default arguments.
+
+```diff
+- def to_bytes(self, *, exclude: List[str] = []):
++ def to_bytes(self, *, exclude: List[str] = SimpleFrozenList()):
+    ...
+```
+
+```diff
+def do_something(values: List[str] = SimpleFrozenList()):
+    if some_condition:
+-         values.append("foo")  # raises an error
++         values = [*values, "foo"]
+    return values
+```
+
+### Don't use `try`/`except` for control flow
+
+We strongly discourage using `try`/`except` blocks for anything that's not third-party error handling or error handling that we otherwise have little control over. There's typically always a way to anticipate the _actual_ problem and **check for it explicitly**, which makes the code easier to follow and understand, and prevents bugs:
+
+```diff
+- try:
+-     token = doc[i]
+- except IndexError:
+-     token = doc[-1]
+
++ if i < len(doc):
++     token = doc[i]
++ else:
++     token = doc[-1]
+```
+
+Even if you end up having to check for multiple conditions explicitly, this is still preferred over a catch-all `try`/`except`. It can be very helpful to think about the exact scenarios you need to cover, and what could go wrong at each step, which often leads to better code and fewer bugs. `try/except` blocks can also easily mask _other_ bugs and problems that raise the same errors you're catching, which is obviously bad.
+
+If you have to use `try`/`except`, make sure to only include what's **absolutely necessary** in the `try` block and define the exception(s) explicitly. Otherwise, you may end up masking very different exceptions caused by other bugs.
+
+```diff
+- try:
+-     value1 = get_some_value()
+-     value2 = get_some_other_value()
+-     score = external_library.compute_some_score(value1, value2)
+- except:
+-     score = 0.0
+
++ value1 = get_some_value()
++ value2 = get_some_other_value()
++ try:
++     score = external_library.compute_some_score(value1, value2)
++ except ValueError:
++     score = 0.0
+```
+
+### Avoid lambda functions
```

Comment:
I think this section is a bit strongly worded. There's still tonnes of times where lambdas are the right thing to do, e.g. when you're passing a key to a sorting function.

The issue is really when you're passing lambdas into somewhere where they might get stored. So people should be a bit careful of using them, but avoiding lambdas entirely could make for some ugly code.

Reply:
Sure, we can phrase this so it sounds less strict. I do think there are very few cases where lambdas are the right choice. Sort keys that are short enough to be readable in one line are the main exception.

Btw, fun fact: The only places in spaCy where we currently use lambdas are serializers, `get_examples` callbacks and tests. Nothing else.